### PR TITLE
drivers: add mutex to the i2c-gpio driver.

### PR DESCRIPTION
**Pull request by yeongnamChu**
Mutexs were added to improve the overlapping phenomenon because there is a part that shares the i2c line.

### CODE REVIEW

**Review by XenuIsWatching — CHANGES_REQUESTED** [review]
This should be a mutex and not a binary semaphore to prevent a priority inversion

**Comment by yeongnamChu** [comment]
> This should be a mutex and not a binary semaphore to prevent a priority inversion

Okay. I will change it to mutex and upload it.

**Review by teburd — APPROVED** [review]

File: drivers/i2c/i2c_gpio.c [hunk 1]
```diff
@@ -47,6 +47,7 @@ struct i2c_gpio_config {
 /* Driver instance data */
 struct i2c_gpio_context {
 	struct i2c_bitbang bitbang;	/* Bit-bang library data */
+	struct k_mutex mutex;
```
Comment (xakep-amatop) [inline drivers/i2c/i2c_gpio.c:50]: as far as I know, mutexes are designed for threads. However, can we be sure that changed functions won't be invoked from an ISR?

**Review by XenuIsWatching — APPROVED** [review]

File: drivers/i2c/i2c_gpio.c (hunk 1 above)
Comment (XenuIsWatching) [reply drivers/i2c/i2c_gpio.c:50]: you shouldn't be calling a 'bit-banged' driver function that is gonna take a long time from the beginning, that would be a very bad ISR as that could take a long time.

Comment (xakep-amatop) [reply drivers/i2c/i2c_gpio.c:50]: so, should we add a note about this somewhere? "shouldn't" -- it is good, but my question was - can we be sure about that?
what about calls with IRQs disabled, e.g. atomic context?

Comment (yeongnamChu) [reply drivers/i2c/i2c_gpio.c:50]: Then, do you mean to prevent IRQs from occurring while using I2C GPIO?

Comment (XenuIsWatching) [reply drivers/i2c/i2c_gpio.c:50]: The mutex api already has asserts that check if it is within an ISR. Anything outside of that is redundant and unnecessary. Every place that I've seen use a muted does not need an additional comment or code check as those should understand that a muted already is. 

The code looks good as is... but you gotta fix the compliance issues with the git commit and push again.

Comment (yeongnamChu) [reply drivers/i2c/i2c_gpio.c:50]: I fixed the compliance issue with git commit and pushed it again.

**Comment by yeongnamChu** [comment]
I fixed the compliance issue with git commit and pushed it again.

File: drivers/i2c/i2c_gpio.c (hunk 1 above)
Comment (xakep-amatop) [reply drivers/i2c/i2c_gpio.c:50]: got it, thank you

**Pull request merged**